### ISSUE TITLE
rack: rake test causes uninitialized constant Rack::MockResponse (NameError)

### ISSUE THREAD

**Issue by deepj** [oracle/truffleruby#1661]
This has started happening since RC15, not in RC14.

To reproduce:

```
git clone https://github.com/rack/rack.git
cd rake
bundle
rake test
```

RC15 + nightly

```
$ rake test
~/.rubies/truffleruby-1.0.0-rc15/bin/truffleruby -I"lib:test" -I"~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib" "~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb" "test/spec_auth_basic.rb" "test/spec_auth_digest.rb" "test/spec_body_proxy.rb" "test/spec_builder.rb" "test/spec_cascade.rb" "test/spec_cgi.rb" "test/spec_chunked.rb" "test/spec_common_logger.rb" "test/spec_conditional_get.rb" "test/spec_config.rb" "test/spec_content_length.rb" "test/spec_content_type.rb" "test/spec_deflater.rb" "test/spec_directory.rb" "test/spec_etag.rb" "test/spec_events.rb" "test/spec_fastcgi.rb" "test/spec_file.rb" "test/spec_handler.rb" "test/spec_head.rb" "test/spec_lint.rb" "test/spec_lobster.rb" "test/spec_lock.rb" "test/spec_logger.rb" "test/spec_media_type.rb" "test/spec_method_override.rb" "test/spec_mime.rb" "test/spec_mock.rb" "test/spec_multipart.rb" "test/spec_null_logger.rb" "test/spec_recursive.rb" "test/spec_request.rb" "test/spec_response.rb" "test/spec_rewindable_input.rb" "test/spec_runtime.rb" "test/spec_sendfile.rb" "test/spec_server.rb" "test/spec_session_abstract_id.rb" "test/spec_session_abstract_session_hash.rb" "test/spec_session_cookie.rb" "test/spec_session_memcache.rb" "test/spec_session_pool.rb" "test/spec_show_exceptions.rb" "test/spec_show_status.rb" "test/spec_static.rb" "test/spec_tempfile_reaper.rb" "test/spec_thin.rb" "test/spec_urlmap.rb" "test/spec_utils.rb" "test/spec_version.rb" "test/spec_webrick.rb" "test/gemloader.rb"
~/.rubies/truffleruby-1.0.0-rc15/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `const_missing': uninitialized constant Rack::MockResponse (NameError)
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
	from ~/dev/bugs/truffleruby/rack/lib/rack/mock.rb:25:in `Rack'
	from ~/dev/bugs/truffleruby/rack/lib/rack/mock.rb:11:in `<top (required)>'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/mri/rubygems/core_ext/kernel_require.rb:54:in `require'
	from ~/dev/bugs/truffleruby/rack/test/spec_auth_basic.rb:6:in `<top (required)>'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:17:in `require'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:17:in `require'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:17:in `block in <main>'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:5:in `select'
	from ~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:5:in `<main>'
rake aborted!
Command failed with status (1): [ruby -I"lib:test" -I"~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib" "~/.rubies/truffleruby-1.0.0-rc15/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb" "test/spec_auth_basic.rb" "test/spec_auth_digest.rb" "test/spec_body_proxy.rb" "test/spec_builder.rb" "test/spec_cascade.rb" "test/spec_cgi.rb" "test/spec_chunked.rb" "test/spec_common_logger.rb" "test/spec_conditional_get.rb" "test/spec_config.rb" "test/spec_content_length.rb" "test/spec_content_type.rb" "test/spec_deflater.rb" "test/spec_directory.rb" "test/spec_etag.rb" "test/spec_events.rb" "test/spec_fastcgi.rb" "test/spec_file.rb" "test/spec_handler.rb" "test/spec_head.rb" "test/spec_lint.rb" "test/spec_lobster.rb" "test/spec_lock.rb" "test/spec_logger.rb" "test/spec_media_type.rb" "test/spec_method_override.rb" "test/spec_mime.rb" "test/spec_mock.rb" "test/spec_multipart.rb" "test/spec_null_logger.rb" "test/spec_recursive.rb" "test/spec_request.rb" "test/spec_response.rb" "test/spec_rewindable_input.rb" "test/spec_runtime.rb" "test/spec_sendfile.rb" "test/spec_server.rb" "test/spec_session_abstract_id.rb" "test/spec_session_abstract_session_hash.rb" "test/spec_session_cookie.rb" "test/spec_session_memcache.rb" "test/spec_session_pool.rb" "test/spec_show_exceptions.rb" "test/spec_show_status.rb" "test/spec_static.rb" "test/spec_tempfile_reaper.rb" "test/spec_thin.rb" "test/spec_urlmap.rb" "test/spec_utils.rb" "test/spec_version.rb" "test/spec_webrick.rb" "test/gemloader.rb" ]

Tasks: TOP => test => test:regular
(See full trace by running task with --trace)
```

RC14

```
$ rake test
~/.rubies/truffleruby-1.0.0-rc14/bin/truffleruby -I"lib:test" -I"~/.rubies/truffleruby-1.0.0-rc14/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib" "~/.rubies/truffleruby-1.0.0-rc14/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb" "test/spec_auth_basic.rb" "test/spec_auth_digest.rb" "test/spec_body_proxy.rb" "test/spec_builder.rb" "test/spec_cascade.rb" "test/spec_cgi.rb" "test/spec_chunked.rb" "test/spec_common_logger.rb" "test/spec_conditional_get.rb" "test/spec_config.rb" "test/spec_content_length.rb" "test/spec_content_type.rb" "test/spec_deflater.rb" "test/spec_directory.rb" "test/spec_etag.rb" "test/spec_events.rb" "test/spec_fastcgi.rb" "test/spec_file.rb" "test/spec_handler.rb" "test/spec_head.rb" "test/spec_lint.rb" "test/spec_lobster.rb" "test/spec_lock.rb" "test/spec_logger.rb" "test/spec_media_type.rb" "test/spec_method_override.rb" "test/spec_mime.rb" "test/spec_mock.rb" "test/spec_multipart.rb" "test/spec_null_logger.rb" "test/spec_recursive.rb" "test/spec_request.rb" "test/spec_response.rb" "test/spec_rewindable_input.rb" "test/spec_runtime.rb" "test/spec_sendfile.rb" "test/spec_server.rb" "test/spec_session_abstract_id.rb" "test/spec_session_abstract_session_hash.rb" "test/spec_session_cookie.rb" "test/spec_session_memcache.rb" "test/spec_session_pool.rb" "test/spec_show_exceptions.rb" "test/spec_show_status.rb" "test/spec_static.rb" "test/spec_tempfile_reaper.rb" "test/spec_thin.rb" "test/spec_urlmap.rb" "test/spec_utils.rb" "test/spec_version.rb" "test/spec_webrick.rb" "test/gemloader.rb"
Lighttpd isn't installed. Skipping Rack::Handler::CGI and FastCGI tests. Install lighttpd to run them.
Skipping Rack::Session::Memcache tests (Memcache is required). `gem install memcache-client` and try again.
Skipping Rack::Handler::Thin tests (Thin is required). `gem install thin` and try again.
Run options: --seed 37340

# Running:

....................................E...........S.....S...........................^Crake aborted!
Interrupt: Interrupt

Tasks: TOP => test => test:regular
(See full trace by running task with --trace)
Interrupted. Exiting...


Finished in 5.612750s, 14.6096 runs/s, 46.8576 assertions/s.
```

**Comment by deepj** [comment]
The same issue when I try to run sprockets tests

```
bundle exec rake test
~/.rubies/truffleruby-1.0.0-nightly/lib/truffle/rbconfig.rb:35: warning: loading in progress, circular require considered harmful - ~/.rubies/truffleruby-1.0.0-nightly/lib/truffle/rbconfig.rb
~/.rubies/truffleruby-1.0.0-nightly/lib/mri/rubygems/requirement.rb:29: warning: loading in progress, circular require considered harmful - ~/.rubies/truffleruby-1.0.0-nightly/lib/mri/rubygems/requirement.rb
~/.rubies/truffleruby-1.0.0-nightly/lib/mri/tmpdir.rb:87: warning: shadowing outer local variable - path
~/.gem/truffleruby/2.6.2/gems/concurrent-ruby-1.1.5/lib/concurrent/thread_safe/util/striped64.rb:100: warning: `*' interpreted as argument prefix
~/.gem/truffleruby/2.6.2/bundler/gems/rack-235802118258/lib/rack/mock.rb:25: warning: loading in progress, circular require considered harmful - ~/.gem/truffleruby/2.6.2/bundler/gems/rack-235802118258/lib/rack/mock.rb
~/.gem/truffleruby/2.6.2/bundler/gems/rack-235802118258/lib/rack/mock.rb:25:in `const_missing': uninitialized constant Rack::MockResponse (NameError)
	from ~/.gem/truffleruby/2.6.2/bundler/gems/rack-235802118258/lib/rack/mock.rb:25:in `require'
	from ~/.gem/truffleruby/2.6.2/bundler/gems/rack-235802118258/lib/rack/mock.rb:25:in `Rack'
	from ~/.gem/truffleruby/2.6.2/bundler/gems/rack-235802118258/lib/rack/mock.rb:11:in `<top (required)>'
	from ~/dev/bugs/truffleruby/sprockets/test/test_environment.rb:3:in `require'
	from ~/dev/bugs/truffleruby/sprockets/test/test_environment.rb:3:in `<top (required)>'
	from ~/.rubies/truffleruby-1.0.0-nightly/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:17:in `require'
	from ~/.rubies/truffleruby-1.0.0-nightly/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:17:in `block in <main>'
	from ~/.rubies/truffleruby-1.0.0-nightly/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:5:in `select'
	from ~/.rubies/truffleruby-1.0.0-nightly/lib/ruby/gems/2.6.0/gems/rake-12.3.2/lib/rake/rake_test_loader.rb:5:in `<main>'
rake aborted!
Command failed with status (1)

Tasks: TOP => test
(See full trace by running task with --trace)
```

**Comment by deepj** [comment]
^^ @chrisseaton @eregon

**Comment by deepj** [comment]
The same issue when I try to run [gel](https://github.com/gel-rb/gel) tests...

**Comment by deepj** [comment]
@eregon Hello, it would be possible take a look at this, please?

**Comment by deepj** [comment]
@chrisseaton @eregon Can you take a look at this? I've seen more problems with `uninitialized constant` kind of problems.

**Comment by eregon** [comment]
OK, I'll to take a look and reproduce.

**Comment by deepj** [comment]
@eregon the same issue has Falcon. See https://travis-ci.com/socketry/falcon/jobs/217858985

**Comment by eregon** [comment]
> @eregon the same issue has Falcon. See https://travis-ci.com/socketry/falcon/jobs/217858985

I think that's unrelated and just caused by the first exception (#1715).

**Comment by rafaelfranca** [comment]
This should be fixed now. Rack test suite is green with the master branch.

**Comment by deepj** [comment]
@rafaelfranca I can't find it in changelog. In what commit it was fixed?

**Comment by eregon** [comment]
@deepj It's not fixed in 19.3, which the 20.0 milestone reflects.
I think https://github.com/oracle/truffleruby/pull/1779 fixed this.

**Comment by rafaelfranca** [comment]
Yes, that is the fix
